qiskit_nature/algorithms/pes_samplers/bopes_sampler.py

```diff
@@ -219,9 +265,78 @@ def _run_single_point(self, point: float) -> EigenstateResult:
+            self._gss.solver, VariationalAlgorithm
+        ):  # type: ignore
             optimal_params = result.raw_result.optimal_point
             self._points_optparams[point] = optimal_params
 
         return result
+
+    def _run_single_point_excited_solver(self, point: float) -> EigenstateResult:
+        """Run the sampler at the given single point
+
+        Args:
+            point: The value of the degree of freedom to evaluate.
+
+        Returns:
+            Results for a single point.
+        Raises:
+            QiskitNatureError: Invalid Driver
+        """
+
+        # update molecule geometry and thus resulting Hamiltonian based on specified point
+
+        if not isinstance(
+            self._driver, (ElectronicStructureMoleculeDriver, VibrationalStructureMoleculeDriver)
+        ):
+            raise QiskitNatureError(
+                "Driver must be ElectronicStructureMoleculeDriver or VibrationalStructureMoleculeDriver."
+            )
+
+        self._driver.molecule.perturbations = [point]
+
+        # find closest previously run point and take optimal parameters
+        if (
+            self._bootstrap
+            and hasattr(self._gss, "_gsc")
+            and isinstance(self._gss._gsc.solver, VariationalAlgorithm)
+        ):  # type: ignore
+            prev_points = list(self._points_optparams.keys())
+            prev_params = list(self._points_optparams.values())
+            n_pp = len(prev_points)
+
+            # set number of points to bootstrap
+            if self._extrapolator is None:
+                n_boot = len(prev_points)  # bootstrap all points
+            else:
+                n_boot = self._num_bootstrap
+
+            # Set initial params # if prev_points not empty
+            if prev_points:
+                if n_pp <= n_boot:
+                    distances = np.array(point) - np.array(prev_points).reshape(n_pp, -1)
+                    # find min 'distance' from point to previous points
+                    min_index = np.argmin(np.linalg.norm(distances, axis=1))
+                    # update initial point
+                    self._gss._gsc.solver.initial_point = prev_params[min_index]  # type: ignore
+                else:  # extrapolate using saved parameters
+                    opt_params = self._points_optparams
+                    param_sets = self._extrapolator.extrapolate(
+                        points=[point], param_dict=opt_params
+                    )
+                    # update initial point, note param_set is a dictionary
+                    self._gss._gsc.solver.initial_point = param_sets.get(point)  # type: ignore
+
+        # the output is an instance of EigenstateResult
+        result = self._gss.solve(self._problem)
+
+        if (
+            isinstance(self._gss, ExcitedStatesSolver)
+            and hasattr(self._gss, "_gsc")
+            and isinstance(self._gss._gsc.solver, VariationalAlgorithm)
+        ):
```

Comment:
I think it would make sense to cache all these checks in another variable. That way you only need to evaluate all three conditions once and can then rely on the boolean result which you stored

Reply:
Still an open question: should we have a local attribute `is_variational = isinstance(self._solver_wrapper.solver, VariationalAlgorithm)` ?